src/vs/workbench/services/textMate/browser/worker/textMateWorkerModel.ts

```diff
@@ -0,0 +1,161 @@
+				const [eolCount] = countEOL(change.text);
+				this._tokenizationStateStore.applyEdits(change.range, eolCount);
+			}
+		}
+		this.tokenizeDebouncer.schedule();
+	}
+
+	public retokenize(startLineNumber: number, endLineNumberExclusive: number) {
+		if (this._tokenizationStateStore) {
+			for (let lineNumber = startLineNumber; lineNumber < endLineNumberExclusive; lineNumber++) {
+				this._tokenizationStateStore.markMustBeTokenized(lineNumber - 1);
+			}
+			this.tokenizeDebouncer.schedule();
+		}
+	}
+
+	private _resetTokenization(): void {
+		this._tokenizationStateStore = null;
+
+		const languageId = this._languageId;
+		const encodedLanguageId = this._encodedLanguageId;
+		this._worker.getOrCreateGrammar(languageId, encodedLanguageId).then((r) => {
+			if (this._isDisposed || languageId !== this._languageId || encodedLanguageId !== this._encodedLanguageId || !r) {
+				return;
+			}
+
+			if (r.grammar) {
+				const tokenizationSupport = new TextMateTokenizationSupport(r.grammar, r.initialState, false);
+				this._tokenizationStateStore = new TokenizationStateStore(tokenizationSupport, tokenizationSupport.getInitialState());
+			} else {
+				this._tokenizationStateStore = null;
+			}
+			this._tokenize();
+		});
+	}
+
+	private _tokenize(): void {
+		if (!this._tokenizationStateStore) {
+			return;
+		}
+
+		const startTime = new Date().getTime();
+
+		while (true) {
+			const builder = new ContiguousMultilineTokensBuilder();
+			const lineCount = this._lines.length;
+
+			let tokenizedLines = 0;
+
+			const stateDeltaBuilder = new StateDeltaBuilder();
+
+			// Validate all states up to and including endLineIndex
+			while (this._tokenizationStateStore.invalidLineStartIndex < lineCount) {
+				const lineIndex = this._tokenizationStateStore.invalidLineStartIndex;
+
+				tokenizedLines++;
+				// TODO don't spam the renderer
+				if (tokenizedLines > 200) {
+					break;
+				}
+
+				const text = this._lines[lineIndex];
+
+				const lineStartState = this._tokenizationStateStore.getBeginState(lineIndex) as StateStack;
+				const tokenizeResult = this._tokenizationStateStore.tokenizationSupport.tokenizeEncoded(text, true, lineStartState);
+				if (this._tokenizationStateStore.setEndState(lineCount, lineIndex, tokenizeResult.endState)) {
+					const delta = diffStateStacksRefEq(lineStartState, tokenizeResult.endState as StateStack);
+					stateDeltaBuilder.setState(lineIndex + 1, delta);
+				}
+
+				LineTokens.convertToEndOffset(tokenizeResult.tokens, text.length);
+				builder.add(lineIndex + 1, tokenizeResult.tokens);
+			}
+
+			if (tokenizedLines === 0) {
+				break;
+			}
+
+			const stateDeltas = stateDeltaBuilder.getStateDeltas();
+			this._worker.setTokensAndStates(this._uri, this._versionId, builder.serialize(), stateDeltas);
+
+			const deltaMs = new Date().getTime() - startTime;
+			if (deltaMs > 20) {
+				// yield to check for changes
+				setTimeout(() => this._tokenize(), 3);
+				break;
+			}
```

Comment:
I think we should move this time check also in the inner while loop from above, otherwise 200 really expensive lines could block the worker thread and prevent yielding for a long time.